### PR TITLE
Airy function cleanup

### LINKED ISSUE #17032: Unclear docstring for airyx
The docstring of `airyx` function reads:

> scaled `k`-th derivative of the Airy function, return `\operatorname{Ai}(x) e^{\frac{2}{3} x \sqrt{x}}` for `k == 0 || k == 1`\ , and `\operatorname{Ai}(x) e^{- \left| \operatorname{Re} \left( \frac{2}{3} x \sqrt{x} \right) \right|}` for `k == 2 || k == 3`.

This is at best unclear, if not misleading.  `airyx(2, x)` isn't the scaled second derivative of Airy Ai function, but rather the scaled Airy Bi function, like `airyx(3, x)` isn't the scaled third derivative of Airy Ai.

In addition, how is the scaling factor computed?  For negative real values of `x` it should be complex according to the docstring (at least in the case `k == 0 || k ==1`), but `airyx(0, -1.23)` is real.  GSL [uses](https://www.gnu.org/software/gsl/manual/html_node/Airy-Functions.html#Airy-Functions) 1 as scaling factor for negative real `x`, but this isn't the case for Julia's `airyx`, because `airyx(0, -1.23) != airy(0, -1.23)`.

### PR DESCRIPTION
**Pull request by simonbyrne**
This rearranges the different airy functions, deprecating `airy`, `airyx` and `airyprime` in favour of the more specific variants (`airyai`, `airyaiprime`, etc.), and clarifies the docs of each remaining function. Fixes #17032.

### CODE REVIEW

File: base/special/bessel.jl
Comment (tkelman) [inline base/special/bessel.jl]: all docstring signature changes need to made in the RST

Comment (simonbyrne) [inline base/special/bessel.jl]: good catch, thanks.

File: doc/stdlib/math.rst
Comment (tkelman) [inline doc/stdlib/math.rst]: and they need to match ... the inline docstring has `::Complex128` here, this one doesn't

Comment (simonbyrne) [inline doc/stdlib/math.rst]: They do: I added `function airyai end` between the docstring and the Complex128 method.

Comment (tkelman) [inline doc/stdlib/math.rst]: the contents don't match https://github.com/JuliaLang/julia/blob/54ec2150afa6ba03344d08a7ab85be89fe265e17/base/special/bessel.jl#L43-L46

Comment (simonbyrne) [inline doc/stdlib/math.rst]: Ah, I missed that one. Thanks

File: base/special/bessel.jl
Comment (ararslan) [inline base/special/bessel.jl]: Should we be deprecating the vectorized versions of these in favor of the dot-broadcast syntax?

**Comment by ararslan** [comment]
This looks great to me! :+1: 

Once this is merged, chance you'd want to apply these same changes in JuliaMath/SpecialFunctions.jl? If not I can do so later.

File: base/special/bessel.jl
Comment (tkelman) [inline base/special/bessel.jl]: if not here, https://github.com/JuliaLang/julia/pull/17302 should

Comment (ararslan) [inline base/special/bessel.jl]: Ah yeah, good call. Looks like the change is already part of that PR.

Comment (simonbyrne) [inline base/special/bessel.jl]: Actually the vectorized forms aren't deprecated; I just moved the `@vectorize_1arg` calls to the `@eval` statement further down.

**Comment by giordano** [comment]
Reading the [code](https://github.com/JuliaLang/julia/blob/a5e7a639cfaf1f9dc6a0ad5ba34f92b164537c25/base/special/bessel.jl#L113) I understand that all the Airy functions of real arguments are always real, is it correct?  So `airyaix(-3) != airyaix(complex(-3))`.  If so, would you mind documenting this?

**Comment by simonbyrne** [comment]
@giordano Thanks, I thought they were, but it turns out not: the Airy functions themselves are real for real values, but the scaled `airyaix` ones are not due to the `sqrt(x)`.

The question here is should we return a complex value, or throw a domain error?

**Comment by simonbyrne** [comment]
@stevengj Your thoughts?

**Comment by simonbyrne** [comment]
Note this is also an issue in the existing code:
```julia
julia> airyx(-3)
0.35928383932620317

julia> airyx(complex(-3))
0.35928383932620317 - 0.1200641157612309im
```

**Comment by stevengj** [comment]
Either throw a domain error, or (similar to `lgamma`) define it to be the analytic complex version for complex `x` and to be scaled by`sqrt(abs(x))` for real `x`.

**Comment by simonbyrne** [comment]
> define it ... to be scaled by `sqrt(abs(x))` for real `x`.

Won't that require changes to the AMOS code?

**Comment by giordano** [comment]
GSL uses as scaling factor the same as Julia for x>0, and 1 for x < 0 (https://www.gnu.org/software/gsl/manual/html_node/Derivatives-of-Airy-Functions.html).  Is this an option for Julia?  Of course it would be backward incompatible.

**Comment by stevengj** [comment]
Looking at this more closely, I would suggest throwing a `DomainError` for negative `Real` arguments.

For negative arguments, there is no exponential growth or decay that needs to be cancelled, so there is no really need for any scale factor.   In that sense, GSL's choice of a scale factor of 1 for `x < 0` makes a certain amount of sense.

However, if you switch scale factors to 1 for negative arguments, then any code *calling* the `airyaix` function would need to take this into account (if it cares about negative `x`).   So, the caller's code will need an `if x < 0` check.  But if they are doing an `if x < 0` check anyway, they can just call `airyai(x)` for `x < 0` themselves — there is no need for us to provide it.

Furthermore, if we silently switch scale factors for `x < 0`, then user code that does *not* check for this case will be incorrect.  It is better to throw a `DomainError` so that they know they need to call the unscaled function manually for the case of negative arguments.

**Comment by simonbyrne** [comment]
I agree with the logic. I've updated the PR and added some more tests for this case.

**Comment by tkelman** [comment]
@simonbyrne any update here?

**Comment by simonbyrne** [comment]
I think the only thing holding this back was the decision on what was happening with #4301/#8536, but we should probably still do it. I'll rebase.

**Comment by tkelman** [comment]
If #8536 happens soon it should probably incorporate this change anyway.

**Comment by simonbyrne** [comment]
Updated.

**Comment by tkelman** [comment]
oops, ignore any travis failures about cmake here, need to restart after https://travis-ci.org/JuliaLang/julia/jobs/186158977 and https://travis-ci.org/JuliaLang/julia/jobs/186158972 have finished and uploaded a new cache

**Review by tkelman — APPROVED** [review]
(squash) merge when green?

**Review by stevengj — APPROVED** [review]
LGTM.

**Comment by giordano** [comment]
Just compiled Julia master, but compatibility function doesn't seem to work:

```julia
julia> airy(0, 4.0)
ERROR: UndefVarError: afn not defined
Stacktrace:
 [1] airy at ./deprecated.jl:1188 [inlined]
 [2] airy(::Int64, ::Float64) at ./deprecated.jl:1196
```

**Comment by tkelman** [comment]
oops, we should sanity check deprecations in a better way...

might be enough to add a dollar sign inside the QuoteNode calls?

**Comment by giordano** [comment]
Do you mean
```julia
$(QuoteNode($afn))
```
?  I tried that but compilation of Julia fails with
```
deprecated.jl
error during bootstrap:
LoadError(at "sysimg.jl" line 369: LoadError(at "deprecated.jl" line 1183: ErrorException("error compiling anonymous: syntax: prefix "$" in non-quoted expression")))
```

**Comment by nalimilan** [comment]
This broke `make docs`:
```
Documenter: setting up build directory.
Documenter: expanding markdown templates.
 !! Undefined binding 'Base.Math.airyaix'. [src/stdlib/math.md]
 !! Undefined binding 'Base.Math.airyaiprimex'. [src/stdlib/math.md]
 !! Undefined binding 'Base.Math.airybix'. [src/stdlib/math.md]
 !! Undefined binding 'Base.Math.airybiprimex'. [src/stdlib/math.md]
Documenter: building cross-references.
 !! No doc found for reference '[`airyaix(z)`](@ref)'. [src/manual/mathematical-operations.md]
 !! No doc found for reference '[`airyaiprimex(z)`](@ref)'. [src/manual/mathematical-operations.md]
 !! No doc found for reference '[`airybix(z)`](@ref)'. [src/manual/mathematical-operations.md]
 !! No doc found for reference '[`airybiprimex(z)`](@ref)'. [src/manual/mathematical-operations.md]
Documenter: running document checks.
 !! Skipped doctesting.
 > checking footnote links.
Documenter: populating indices.
ERROR: LoadError: `makedocs` encountered errors. Terminating build
 in runner(::Type{Documenter.Builder.RenderDocument}, ::Documenter.Documents.Document) at /home/milan/Dev/julia-packaging/julia-master/doc/deps/v0.6/Documenter/src/Builder.jl:202
```

Maybe we should run it on CI to catch this kind of bug?

**Comment by tkelman** [comment]
It did run on CI though? https://travis-ci.org/JuliaLang/julia/jobs/187530248#L745 just didn't flag an error?

**Comment by nalimilan** [comment]
Sorry, probably a bug on my end then. I guess I used the incorrect Julia executable to build the docs.

**Comment by giordano** [comment]
The fact that `airyaix(x)` for negative real `x` throws a `DomainError` should be documented?  Sorry for being pedantic :-)

**Comment by simonbyrne** [comment]
Yes, it should. Would you mind making a pull request?

**Comment by giordano** [comment]
No, I can do it!

**Pull request merged**
